htlcswitch/circuit_map.go

```diff
@@ -0,0 +1,848 @@
+	// exact circuit to be forwarded through the switch. If a circuit is
+	// already pending, the htlc will be reforwarded by the switch.
+	cm.mtx.Lock()
+	var adds, drops, fails, addFails []*PaymentCircuit
+	for _, circuit := range circuits {
+		inKey := circuit.InKey()
+		if foundCircuit, ok := cm.pending[inKey]; ok {
+			switch {
+
+			// This circuit has a keystone, it's waiting for a
+			// response from the remote peer on the outgoing link.
+			// Drop it like it's hot, ensure duplicates get caught.
+			case foundCircuit.HasKeystone():
+				drops = append(drops, circuit)
+
+			// If no keystone is set and the switch has not been
+			// restarted, the corresponding packet should still be
+			// in the outgoing link's mailbox. It will be delivered
+			// if it comes online before the switch goes down.
+			//
+			// NOTE: Dropping here prevents a flapping, incoming
+			// link from failing a duplicate add while it is still
+			// in the server's memory mailboxes.
+			case !foundCircuit.LoadedFromDisk:
+				drops = append(drops, circuit)
+
+			// Otherwise, the in-mem packet has been lost due to a
+			// restart. It is now safe to send back a failure along
+			// the incoming link. The incoming link should be able
+			// detect and ignore duplicate packets of this type.
+			default:
+				fails = append(fails, circuit)
+				addFails = append(addFails, circuit)
+			}
+
+			continue
+		}
+
+		cm.pending[inKey] = circuit
+		adds = append(adds, circuit)
+		addFails = append(addFails, circuit)
+	}
+	cm.mtx.Unlock()
+
+	// If all circuits are dropped or failed, we are done.
+	if len(adds) == 0 {
+		actions.Drops = drops
+		actions.Fails = fails
+		return actions, nil
+	}
+
+	// Now, optimistically serialize the circuits to add.
+	var bs = make([]bytes.Buffer, len(adds))
+	for i, circuit := range adds {
+		if err := circuit.Encode(&bs[i]); err != nil {
+			actions.Drops = drops
+			actions.Fails = addFails
+			return actions, err
+		}
+	}
+
+	// Write the entire batch of circuits to the persistent circuit bucket
+	// using bolt's Batch write. This method must be called from multiple,
+	// distinct goroutines to have any impact on performance.
+	err := cm.db.Batch(func(tx *bolt.Tx) error {
+		circuitBkt := tx.Bucket(circuitAddKey)
+		if circuitBkt == nil {
+			return ErrCorruptedCircuitMap
+		}
+
+		for i, circuit := range adds {
+			inKeyBytes := circuit.InKey().Bytes()
+			circuitBytes := bs[i].Bytes()
+
+			err := circuitBkt.Put(inKeyBytes, circuitBytes)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+
+	// Return if the write succeeded.
+	if err == nil {
+		actions.Adds = adds
+		actions.Drops = drops
+		actions.Fails = fails
+		return actions, nil
+	}
+
+	// Otherwise, rollback the circuits added to the pending set if the
+	// write failed.
+	cm.mtx.Lock()
+	for _, circuit := range adds {
+		inKey := circuit.InKey()
+		if _, ok := cm.pending[inKey]; !ok {
+			delete(cm.pending, inKey)
```

Comment:
I don't see the effect of this

Reply:
Just ensures that our in-memory state is always consistent with on-disk contents, even in face of failures

Reply:
I mean, you check if `inKey` is in the map, and if it is _NOT_, then you attempt to delete it...?

Should it be `ok`? (and in that case, you can just call `delete` without checking if it's in the map first)

Reply:
Oh now I see! Went right over my head the first time. Lemme correct that